src/standard/expr.mli

```diff
@@ -16,736 +16,99 @@ type hash = private int
 type index = private int
 type 'a tag = 'a Tag.t
 
-type let_kind =
-  | Parallel
-  | Sequential
+(** {3 Type definitions} *)
 
-type builtin = ..
-(* Extensible variant type for builtin operations. Encodes in its type
-   arguments the lengths of the expected ty and term arguments respectively. *)
+type builtin = < ty_cst : ty_cst ; term_cst : term_cst > Builtin.t
+(* Extensible variant type for builtin operations. *)
 
-type ttype = Type
-(** The type of types. *)
-
-type 'ty id = private {
-  ty            : 'ty;
+and 'ty id = private {
+  id_ty         : 'ty;
+  index         : index; (** unique index *)
   name          : string;
-  index         : index; (** unique *)
   builtin       : builtin;
   mutable tags  : Tag.map;
 }
 (** The type of identifiers. ['ty] is the type for representing the type of
-    the id, ['ty] and ['t_n] are the lengths of arguments as described by
-    the {builtin} type. *)
+    the id. *)
 
-type ('ttype, 'ty) function_type = {
-  fun_vars : 'ttype id list; (* prenex forall *)
-  fun_args : 'ty list;
-  fun_ret : 'ty;
-}
-(** The type for representing function types. *)
+and type_ = Type
 
+and type_fun = {
+  arity : int;
+  mutable alias : type_alias;
+}
 
-(** {3 Types} *)
+and type_alias =
+  | No_alias
+  | Alias of {
+      alias_vars : ty_var list;
+      alias_body : ty;
+    }
```

Comment:
It's not strictly related to higher-order, but since I'm planning to work on printing back problems right after, keeping type aliases rather than expanding them during type-checking will allow to be able to read some smtlib problem containing type aliases, and then print it back, including the type aliases.